cpp/src/neighbors/detail/hnsw.hpp

```diff
@@ -206,6 +206,213 @@ std::enable_if_t<hierarchy == HnswHierarchy::CPU, std::unique_ptr<index<T>>> fro
+    auto end_idx       = offsets[hist.size() - 1];
+    auto num_pts       = end_idx - start_idx;
+    auto neighbor_size = num_pts > appr_algo->M_ ? appr_algo->M_ : num_pts;
+    if (neighbor_size < 2) {
+      // this means only 1 point in the level
+      continue;
+    }
+
+    // gather points from dataset to form query set on device
+    auto query_set =
+      raft::make_device_matrix<T, int64_t>(res, num_pts, host_dataset_view.extent(1));
+    auto query_set_view = raft::make_device_matrix_view<const T, int64_t>(
+      query_set.data_handle(), query_set.extent(0), query_set.extent(1));
+    auto host_query_set =
+      raft::make_host_matrix<T, int64_t>(query_set.extent(0), query_set.extent(1));
+#pragma omp parallel for
+    for (auto i = start_idx; i < end_idx; i++) {
+      auto pt_id = order[i];
+      std::copy(host_dataset_view.data_handle() + pt_id * host_dataset_view.extent(1),
+                host_dataset_view.data_handle() + (pt_id + 1) * host_dataset_view.extent(1),
+                host_query_set.data_handle() + (i - start_idx) * host_dataset_view.extent(1));
+    }
+    raft::copy(query_set.data_handle(),
+               host_query_set.data_handle(),
+               query_set.size(),
+               raft::resource::get_cuda_stream(res));
+
+    // initialize brute force index
+    auto bf_index = cuvs::neighbors::brute_force::build(
+      res, cuvs::neighbors::brute_force::index_params{}, query_set_view);
+
+    // search for nearest neighbors
+    auto neighbors = raft::make_device_matrix<int64_t, int64_t>(res, num_pts, neighbor_size);
+    auto distances = raft::make_device_matrix<float, int64_t>(res, num_pts, neighbor_size);
+    cuvs::neighbors::brute_force::search(res,
+                                         cuvs::neighbors::brute_force::search_params{},
+                                         bf_index,
+                                         query_set_view,
+                                         neighbors.view(),
+                                         distances.view());
+
+    auto host_neighbors_full =
+      raft::make_host_matrix<int64_t, int64_t>(neighbors.extent(0), neighbors.extent(1));
+    raft::copy(host_neighbors_full.data_handle(),
+               neighbors.data_handle(),
+               neighbors.size(),
+               raft::resource::get_cuda_stream(res));
+    raft::resource::sync_stream(res);
+
+    // Need to slice the first column of the neighbors matrix because it's the query point itself
+    auto host_neighbors =
+      raft::make_host_matrix<int64_t, int64_t>(neighbors.extent(0), neighbors.extent(1) - 1);
+#pragma omp parallel for
+    for (int64_t i = 0; i < host_neighbors.extent(0); i++) {
+      std::copy(host_neighbors_full.data_handle() + i * host_neighbors_full.extent(1) + 1,
+                host_neighbors_full.data_handle() + (i + 1) * host_neighbors_full.extent(1),
+                host_neighbors.data_handle() + i * host_neighbors.extent(1));
```

Comment:
We can use the indexing operator to simplify this:
```suggestion
      std::copy(&host_neighbors(i, 1),   &host_neighbors_full(i + 1, 0),  &host_neighbors(i, 0));
```

But is this really needed to create a copy without the first element? Can't we just skip it in the subsequent loop and save memory?

Reply:
This is not needed anymore with NN Descent